pandas/core/arrays/masked.py

```diff
@@ -280,10 +287,27 @@ def to_numpy(  # type: ignore[override]
         if na_value is lib.no_default:
             na_value = libmissing.NA
         if dtype is None:
-            # error: Incompatible types in assignment (expression has type
-            # "Type[object]", variable has type "Union[str, dtype[Any], None]")
-            dtype = object  # type: ignore[assignment]
-        if self._hasna:
+            if is_numeric_dtype(self):
+                dtype = self.dtype.numpy_dtype
+            else:
+                # error: Incompatible types in assignment (expression has type
+                # "Type[object]", variable has type "Union[str, dtype[Any], None]")
+                dtype = object  # type: ignore[assignment]
+
+        if is_numeric_dtype(self):
+
+            # If there is NA and the data is of int type, a float
+            # is being returned as int type cannot support np.nan.
+            if is_integer_dtype(self) and self._hasna:
+                data = self._data.astype(float)
+            else:
+                data = self._data.astype(dtype)
+
+            # For numerical input, pd.na is replaced with np.nan
+            if self._hasna is True:
+                data[np.where(self._mask is True)] = np.nan
```

Comment:
This line looks suspect ... `self._mask` is an ndarray, so won't comparing identity with True just always give False?

Reply:
Also, what about the specified na_value?